v2/pkg/archive/strict_adder.go

```diff
@@ -0,0 +1,212 @@
+package archive
+
+import (
+	"archive/tar"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	clog "github.com/openshift/oc-mirror/v2/pkg/log"
+)
+
+type strictAdder struct {
+	destination        string
+	archiveFile        *os.File
+	tarWriter          *tar.Writer
+	maxArchiveSize     int64
+	currentChunkId     int
+	sizeOfCurrentChunk int64
+	logger             clog.PluggableLoggerInterface
+}
+
+// `newStrictAdder` initializes the strictAdder implementation for the `archiveAdder` interface.
+// This implementation doesn't allow for any files to exceed the maxArchiveSize specified in the
+// imageSetConfig. It stops adding to the archive chunks if a file exceeds maxArchiveSize
+// and returns in error.
+func newStrictAdder(maxSize int64, destination string, logger clog.PluggableLoggerInterface) (*strictAdder, error) {
+	chunk := 1
+	archiveFileName := fmt.Sprintf("%s_%06d.tar", archiveFilePrefix, chunk)
+	err := os.MkdirAll(destination, 0755)
+	if err != nil {
+		return &strictAdder{}, err
+	}
+	archivePath := filepath.Join(destination, archiveFileName)
+	// Create a new tar archive file
+	// to be closed by BuildArchive
+	archiveFile, err := os.Create(archivePath)
+	if err != nil {
+		return &strictAdder{}, err
+	}
+	// Create a new tar writer
+	// to be closed by the call to close method
+	tarWriter := tar.NewWriter(archiveFile)
+	if maxSize == 0 {
+		maxSize = defaultSegSize * segMultiplier
+	}
+	p := strictAdder{
+		maxArchiveSize:     maxSize,
+		currentChunkId:     chunk,
+		sizeOfCurrentChunk: int64(0),
+		destination:        destination,
+		archiveFile:        archiveFile,
+		tarWriter:          tarWriter,
+		logger:             logger,
+	}
+	return &p, nil
+}
+
+func (o *strictAdder) close() error {
+	err := o.tarWriter.Flush()
+	if err != nil {
+		o.logger.Warn("error flushing archive writer : %v", err)
+	}
+	err = o.tarWriter.Close()
+	if err != nil {
+		o.logger.Warn("error closing archive writer : %v", err)
+	}
+	return o.archiveFile.Close()
+}
+
+// addFile copies the contents of the `pathToFile` file from the disk into
+// the current chunk archive at `pathInTar` location.
+// addFile monitors the size of the current chunk, and creates a new chunk to
+// archive `pathToFile` if archiving would make the chunk exceed the maxArchiveSize.
+// addFile stops the archiving and returns an error if `pathToFile`'s size is
+// greater than maxArchiveSize.
+func (o *strictAdder) addFile(pathToFile string, pathInTar string) error {
+	fi, err := os.Stat(pathToFile)
+	if err != nil {
+		return err
+	}
+
+	// when a file is already bigger than the maxArchiveSize, it will not fit any chunk
+	// therefore we should stop
+	if fi.Size() > o.maxArchiveSize {
```

Comment:
Thank you! great advice. That code was duplicated 4 times in total.
I refactored most of the code from addFile and addAllFolder into addFileToWriter function.
I left out the logic that calls o.nextChunk() and increments the sizeOfCurrentChunk, because these are 2 different types...